### ISSUE TITLE
Problem preparing gems with suffixes like preview or alpha in the version

### ISSUE THREAD

**Issue by robertpanzer** [jruby-gradle/jruby-gradle-plugin#181]
In the current build of AsciidoctorJ we use several plugins that have no final version number yet, like 'ruby gems:asciidoctor-pdf:1.5.0.alpha9'.

I am currently trying to migrate from ruby-gradle.base 0.1.17 to 1.0.1 and now these gems are no longer found.
To demonstrate my problem I created a small example:
```groovy
buildscript {
  repositories {
    jcenter()
  }
  dependencies {
    classpath "com.github.jruby-gradle:jruby-gradle-plugin:1.0.1"
  }
}

apply plugin: 'com.github.jruby-gradle.base'
apply plugin: 'java'

ext {
  jrubyVersion = '9.0.0.0'
}

jruby {
  defaultRepositories = false
  defaultVersion = jrubyVersion
  execVersion = jrubyVersion
}

repositories {
    mavenCentral()
    maven {
      name 'rubygems-release'
      url 'http://rubygems-proxy.torquebox.org/releases'
    }
    maven {
      name 'rubygems-prerelease'
      url 'http://rubygems-proxy.torquebox.org/prereleases'
    }
}

dependencies {
  compile "org.jruby:jruby-complete:$jrubyVersion"
  //gems 'rubygems:asciidoctor:1.5.2'
  gems 'rubygems:asciidoctor-diagram:1.3.0.preview.4'
}
```

While `gradle jrubyPrepare` fails like this, it successfully resolves the dependency on `asciidoctor:1.5.2`.
With Gradle 2.5 I see this error message:
```
$ ./gradlew jrubyPrepare -i
Connected to daemon DaemonInfo{pid=78559, address=[eb646ccd-d004-4850-a07e-a6efb5f97758 port:51058, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], idle=true, context=DefaultDaemonContext[uid=8ab4e925-a419-4065-b6f2-75a5ad5ea1bf,javaHome=/Library/Java/JavaVirtualMachines/jdk1.8.0_40.jdk/Contents/Home,daemonRegistryDir=/Users/robertpanzer/.gradle/daemon,pid=78559,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=DE,-Duser.language=de,-Duser.variant]}. Dispatching request Build{id=54be8c3c-76d8-4874-8749-c3d99a6fb849.1, currentDir=/Users/robertpanzer/dev/rubygradletest}.
The client will now receive all logging from the daemon (pid: 78559). The daemon log file: /Users/robertpanzer/.gradle/daemon/2.5/daemon-78559.out.log
Starting 6th build in daemon [uptime: 4 mins 7.316 secs, performance: 98%, memory: 46% of 954,7 MB]
Executing build with daemon context: DefaultDaemonContext[uid=8ab4e925-a419-4065-b6f2-75a5ad5ea1bf,javaHome=/Library/Java/JavaVirtualMachines/jdk1.8.0_40.jdk/Contents/Home,daemonRegistryDir=/Users/robertpanzer/.gradle/daemon,pid=78559,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=DE,-Duser.language=de,-Duser.variant]
Starting Build
Settings evaluated using settings file '/Users/robertpanzer/dev/rubygradletest/settings.gradle'.
Projects loaded. Root project using build file '/Users/robertpanzer/dev/rubygradletest/build.gradle'.
Included projects: [root project 'rubygradletest']
Evaluating root project 'rubygradletest' using build file '/Users/robertpanzer/dev/rubygradletest/build.gradle'.
All projects evaluated.
Selected primary task 'jrubyPrepare' from project :
Tasks to be executed: [task ':jrubyPrepare']
In-memory cache of /Users/robertpanzer/dev/rubygradletest/.gradle/2.5/taskArtifacts/fileHashes.bin: Size{158}, CacheStats{hitCount=0, missCount=0, loadSuccessCount=0, loadExceptionCount=0, totalLoadTime=0, evictionCount=0}
In-memory cache of /Users/robertpanzer/dev/rubygradletest/.gradle/2.5/taskArtifacts/outputFileStates.bin: Size{1}, CacheStats{hitCount=0, missCount=0, loadSuccessCount=0, loadExceptionCount=0, totalLoadTime=0, evictionCount=0}
In-memory cache of /Users/robertpanzer/dev/rubygradletest/.gradle/2.5/taskArtifacts/fileSnapshots.bin: Size{2}, CacheStats{hitCount=0, missCount=0, loadSuccessCount=0, loadExceptionCount=0, totalLoadTime=0, evictionCount=0}
In-memory cache of /Users/robertpanzer/dev/rubygradletest/.gradle/2.5/taskArtifacts/taskArtifacts.bin: Size{1}, CacheStats{hitCount=0, missCount=0, loadSuccessCount=0, loadExceptionCount=0, totalLoadTime=0, evictionCount=0}
:jrubyPrepare (Thread[Daemon worker Thread 3,5,main]) started.
:jrubyPrepare
Resource missing. [HTTP GET: https://repo1.maven.org/maven2/rubygems/asciidoctor-diagram/maven-metadata.xml]
Resource missing. [HTTP GET: https://repo1.maven.org/maven2/rubygems/asciidoctor-diagram/]
:jrubyPrepare (Thread[Daemon worker Thread 3,5,main]) completed. Took 1.528 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all dependencies for configuration ':gems'.
> Could not find any version that matches rubygems:asciidoctor-diagram:[1.3.0.preview.4,1.3.0.preview.4].
  Versions that do not match:
      1.3.0
      1.2.1
      1.2.0
      1.1.6
      1.1.5
      + 13 more
  Required by:
      :rubygradletest:unspecified

* Try:
Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output.

BUILD FAILED
```

I am unsure if I have made sth wrong, but I think I am staying with the documented way: http://jruby-gradle.org/base/#using-a-custom-gem-repository

**Comment by rtyler** [comment]
@robertpanzer In the 1.x of JRuby/Gradle we've switched from relying on the `rubygems.torquebox.org` proxy to something that's more within our control that I've deployed to AWS: rubygems.lasagna.io.

Part of this is for performance (lasagna.io is 100% auto-scaled) but the major part of it is that the torquebox proxy is running a very old version of [rubygems-servlet](https://github.com/torquebox/rubygems-servlet). Whereas `rubygems.lasagna.io` runs the latest and will be updated regularly when changes are available.


I cannot find a pre-releases repo that is valid on the TB proxy, but [rubygems.lasagna.io/proxy/maven/prereleases](http://rubygems.lasagna.io/proxy/maven/prereleases) is definitely functioning correctly and if I fetch the [asciidoctor-pdf maven-metadata.xml](http://rubygems.lasagna.io/proxy/maven/prereleases/rubygems/asciidoctor-pdf/maven-metadata.xml) I see the alpha releases. Although, I believe the rubygems-servlet code is appending `-SNAPSHOT` to them to make it match up more with what one can expect in the Java community.

**Comment by rtyler** [comment]
The long term fix for this is #180 of course :)

**Comment by robertpanzer** [comment]
Hmm.. it should actually work without the prereleases repo.
I can get the following URL without any problems:
http://rubygems-proxy.torquebox.org/releases/rubygems/asciidoctor-diagram/1.3.0.preview.4/asciidoctor-diagram-1.3.0.preview.4.pom
and I can even get it via the lasagna:
http://rubygems.lasagna.io/proxy/maven/releases/rubygems/asciidoctor-diagram/1.3.0.preview.4/asciidoctor-diagram-1.3.0.preview.4.pom
But it's not in the maven-metadata.xml: http://rubygems.lasagna.io/proxy/maven/releases/rubygems/asciidoctor-diagram/maven-metadata.xml
and not in the maven-metadata.xml of torquebox either.

But I don't really understand why it uses the lasagna when I configure `defaultRepositories=false`?

**Comment by mkristian** [comment]
if the prereleases would show up
in releases/rubygems/asciidoctor-diagram/maven-metadata.xml then you will
get prereleases for all version ranges. you can get prereleased
gem-artifact via the release one mainly some gems do reference prerelease
version, or you can use picked a gem with prerelease version in your
project without going prerelease for all gems.

some is true for the torquebox proxy as well the lasagne-proxy

**Comment by robertpanzer** [comment]
OK, next try ;-), sorry for adding the prerelease repo 

`gradle jrubyPrepare --refresh-dependencies` works with this (The old version of the plugin, 0.1.17) :
```groovy
buildscript {
  repositories {
    jcenter()
  }
  dependencies {
    classpath "com.github.jruby-gradle:jruby-gradle-plugin:0.1.17"
  }
}

apply plugin: 'com.github.jruby-gradle.base'
apply plugin: 'java'

ext {
  jrubyVersion = '9.0.0.0'
}

jruby {
  defaultRepositories = false
  defaultVersion = jrubyVersion
  execVersion = jrubyVersion
}

repositories {
    mavenCentral()
    maven {
      name 'rubygems-release'
      url 'http://rubygems-proxy.torquebox.org/releases'
    }
}

dependencies {
  compile "org.jruby:jruby-complete:$jrubyVersion"
  gems 'rubygems:asciidoctor-diagram:1.3.0.preview.4'
}
```

but not with this:
```groovy
buildscript {
  repositories {
    jcenter()
  }
  dependencies {
    classpath "com.github.jruby-gradle:jruby-gradle-plugin:1.0.1"
  }
}

apply plugin: 'com.github.jruby-gradle.base'
apply plugin: 'java'

ext {
  jrubyVersion = '9.0.0.0'
}

jruby {
  defaultRepositories = false
  defaultVersion = jrubyVersion
  execVersion = jrubyVersion
}

repositories {
    mavenCentral()
    maven {
      name 'rubygems-release'
      url 'http://rubygems-proxy.torquebox.org/releases'
    }
}

dependencies {
  compile "org.jruby:jruby-complete:$jrubyVersion"
  gems 'rubygems:asciidoctor-diagram:1.3.0.preview.4'
}
```

**Comment by rtyler** [comment]
The first example doesn't actually work for me on the first go (Gradle 2.4).

```
➜  jruby-gradle-181  gradle jrubyPrepare
:jrubyPrepareGems
Download http://rubygems-proxy.torquebox.org/releases/rubygems/asciidoctor-diagram/1.3.0.preview.4/asciidoctor-diagram-1.3.0.preview.4.pom
Download http://rubygems-proxy.torquebox.org/releases/rubygems/asciidoctor-diagram/1.3.0.preview.4/asciidoctor-diagram-1.3.0.preview.4.gem
Successfully installed asciidoctor-diagram-1.3.0.preview.4
Successfully installed asciidoctor-1.5.2
2 gems installed
:jrubyPrepareGems FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jrubyPrepareGems'.
> Cannot get property 'major' on null object

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

Comically enough, it succeeds on the second run.

The second example fails for me with this:

```
➜  jruby-gradle-181  gradle jrubyPrepare
:jrubyPrepare

FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all dependencies for configuration ':gems'.
> Could not find any version that matches rubygems:asciidoctor-diagram:[1.3.0.preview.4,1.3.0.preview.4].
  Versions that do not match:
      1.3.0
      1.2.1
      1.2.0
      1.1.6
      1.1.5
      + 5 more
  Required by:
      :jruby-gradle-181:unspecified

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

This makes sense to me considering what is in the [maven-metadata.xml](http://rubygems.lasagna.io/proxy/maven/prereleases/rubygems/asciidoctor-diagram/maven-metadata.xml). If I add the prereleases repo, and change the declaration in the `build.gradle` to `gems 'rubygems:asciidoctor-diagram:1.3.0.preview.4-SNAPSHOT' then this works:

```
➜  jruby-gradle-181  gradle jrubyPrepare
:jrubyPrepare
Download http://rubygems.lasagna.io/proxy/maven/prereleases/rubygems/asciidoctor-diagram/1.3.0.preview.4-SNAPSHOT/asciidoctor-diagram-1.3.0.preview.4-SNAPSHOT.pom
Download http://rubygems.lasagna.io/proxy/maven/prereleases/rubygems/asciidoctor-diagram/1.3.0.preview.4-SNAPSHOT/asciidoctor-diagram-1.3.0.preview.4-SNAPSHOT.gem
Successfully installed asciidoctor-diagram-1.3.0.preview.4
Successfully installed asciidoctor-1.5.2
2 gems installed

BUILD SUCCESSFUL

Total time: 28.025 secs

This build could be faster, please consider using the Gradle Daemon: http://gradle.org/docs/2.4/userguide/gradle_daemon.html
➜  jruby-gradle-181  
```


----

@mkristian I think the automatic "-SNAPSHOT" suffix might be pretty confusing for developers, I question whether it's valuable to keep around (if so I'll document it thoroughly) otherwise we could remove it.


Will file another issue to capture "working with pre-releases"

**Comment by mkristian** [comment]
@rtyler no adding the prerelease repo is a bad idea since the version resolution of the transitive dependencies will look for release AND prereleased gems.

FYI the second run on TB repo worked because it takes its time to "download" the gems to produce the all the xml for this gem. once this xml is cached things are fast enough :) lasagne suffers the same performance penalty for first time gems

BUT gradle NOT picking this
```
Could not find any version that matches rubygems:asciidoctor-diagram:[1.3.0.preview.4,1.3.0.preview.4].
```
is a bug on our gradle dependencies resolution. should not be a range but ```rubygems:asciidoctor-diagram:1.3.0.preview.4``` - at least maven will not use the maven-metadata.xml for those artifacts.

in short prereleased versions are meant to work with release repo when picked via a fixed version.
